### ISSUE TITLE
Move pathPrefix from `AbstractAdapter` to a trait.

### ISSUE THREAD

**Issue by SamMousa** [thephpleague/flysystem#798]
It makes sense to move the functionality in `AbstractAdapter` to a trait.
That way other adapters that do not extend from `AbstractAdapter` can still use it.

I can make a PR if you agree with the change.

**Comment by frankdejonge** [comment]
That could work. Traits are supposed to be stateless, but we use PHP so ... YOLO. As long as the tests fully cover the extraction in isolation. You can do this by creating a test subject which only uses the trait and testing that.

**Comment by SamMousa** [comment]
Thinking about it, would it not make even more sense to do path prefixing in a decorator?
(I realize that would break backward compatibility)

That way adapters don't even have to worry about it (by calling `applyPathPrefix` everywhere).

`PrefixAdapter::__construct(AdapterInterface $a, string $prefix)`

**Comment by frankdejonge** [comment]
Theoretically it kind of would, but there's no real benefit. Those kind of changes makes sense when you're dealing with too much complexity, but we're not. I don't like to make breaking API changes for theoretical benefits. I'd basically be putting people to work without any real-life benefits.

**Comment by SamMousa** [comment]
That's true for refactoring your existing core adapters yes, but that is not an argument against including the prefix adapter.
My use case: I use `PDOAdapter` for a proof of concept.
https://github.com/IntegralSoftware/flysystem-pdo-adapter

It does not extend `AbstractAdapter`, I now have to ask 1 of packages to change their stuff, or have to manually extend the `PDOAdapter` and copy paste the path prefix code.

If there was a decorator I could just use that instead.

Basically I could:
- Move the path prefix functionality to a trait (this is code cleanup).
- Include the trait in `AbstractAdapter` (to maintain backward compatibility).
- Create a prefix adapter that uses the trait and proxies requests to the decorated object.

**Comment by frankdejonge** [comment]
I'd be open for a PR for an additional PrefixDecorator (in addition to the trait) but not for one that removes the functionality from the AbstractAdapter.

**Comment by SamMousa** [comment]
Okay, working on that now, some questions:
1. Class name for the decorator? Things like `CachedAdapter` use adapter not decorator. I propose `PrefixAdapter`
2. Does this new adapter extend `AbstractAdapter`. Current core adapters do, but I do not think we should do that for any new adapters, it doesn't add anything. Also separately packaged adapters don't do this either.
My proposal: `PrefixAdapter implements AdapterInterface`

**Comment by frankdejonge** [comment]
I'd call it `PathPrefixAdapter`, it would indeed only implement the `AdapterInterface`, and won't need to extend the `AbstractAdapter`.

**Comment by SamMousa** [comment]
The PR is here: #799, I need some feedback on it though :)

**Comment by twistor** [comment]
Conversation happening in the PR.
